sopt/src/main/scala/dagr/sopt/ArgTokenizer.scala

```diff
@@ -69,19 +74,34 @@ class ArgTokenizer(args: TraversableOnce[String]) extends Iterator[Try[ArgTokeni
   }
 
   /** Returns any remaining args that were not tokenized. This is a destructive operation, so should only be called once. */
-  def takeRemaining: Traversable[String] = this.iterator.toTraversable
+  def takeRemaining: Traversable[String] = this.stack.toTraversable
 
   /** Sets the `nextToken` if it is not defined and we have more strings in the iterator */
-  private def updateNextToken(): Unit = {
-    if (nextToken.isEmpty && iterator.hasNext) {
-      nextToken = iterator.next() match {
-        case "--" => None
-        case ""   => Some(Failure(new OptionNameException("Empty argument given.")))
-        case arg if arg.startsWith("--") => Some(convertDoubleDashOption(arg.substring(2)))
-        case arg if arg.startsWith("-") => Some(convertSingleDash(arg.substring(1)))
-        case arg => Some(Success(ArgValue(value = arg)))
+  private def updateNextToken(): Unit = if (nextToken.isEmpty && this.stack.nonEmpty) nextToken = takeNextToken
+
+  /**
+    * Requires that the stack have at least one item left in it; pulls the top item from the
+    * stack and processes it.
+    */
+  private def takeNextToken: Option[Try[Token]] = (this.stack.pop(), argFilePrefix) match {
+    case ("--", _) => None
+    case ("",   _) => Some(Failure(new OptionNameException("Empty argument given.")))
+    case (arg,  _) if arg.startsWith("--") => Some(convertDoubleDashOption(arg.substring(2)))
+    case (arg,  _) if arg.startsWith("-") => Some(convertSingleDash(arg.substring(1)))
+    case (arg, Some(pre)) if arg.startsWith(pre) =>
+      try {
+        this.stack.pushAll(loadArgumentFile(arg.drop(pre.length)).reverseIterator)
+        if (this.stack.nonEmpty) takeNextToken else None
       }
-    }
+      catch {
+        case ex: IOException => Some(Failure(ex))
+      }
```

Comment:
Thanks.  I think I was trying to be _too_ clever, and didn't think of just doing a secondary match.  I ended up with a modified version of your second option so that I can do this:

```scala
// Now returns Try
private def loadArgumentFile(filename: String): Try[Seq[String]] = {
  Try { Files.readAllLines(Paths.get(filename)).map(s => s.trim).filter(s => s.nonEmpty).toSeq }
}

// That bit of the match
case (arg, Some(pre)) if arg.startsWith(pre) =>
  loadArgumentFile(arg.drop(pre.length)) match {
    case Failure(failure) => Some(Failure(failure))
    case Success(newArgs) =>
      this.stack.pushAll(newArgs.reverseIterator)
      if (this.stack.nonEmpty) takeNextToken else None
  }